### PR TITLE
Expose timestamp field type on coordinator node

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today a coordinating node does not have (easy) access to the mappings
for the indices for the searches it wishes to coordinate. This means it
can't properly interpret a timestamp range filter in a query and must
involve a copy of every shard in at least the `can_match` phase. It
therefore cannot cope with cases when shards are temporarily not started
even if those shards are irrelevant to the search.

This commit captures the mapping of the `@timestamp` field for indices
which expose a timestamp range in their index metadata.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+     * @return the field type of the {@code @timestamp} field of the given index, or {@code null} if:
+     * - the index is not found,
+     * - the field is not found,
+     * - the mapping is not known yet, or
+     * - the field is not a timestamp field.
+     */
+    @Nullable
+    public DateFieldMapper.DateFieldType getTimestampFieldType(Index index) {
+        final PlainActionFuture<DateFieldMapper.DateFieldType> future = fieldTypesByIndex.get(index);
+        if (future == null || future.isDone() == false) {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java:168]: There remains a question of whether we should block here or not (and if so, for how long).

On reflection I think we shouldn't block. Returning `null` sooner will allow the search coordination to proceed normally, ignoring any timestamp filter and deferring any skipping to the individual shards. This means we'll see shard failures if the coordinating node falls behind on extracting these mappings AND some of the shards are unassigned, which is hopefully rare.

As a follow-up we could in theory add another more patient getter to support a workflow that goes:

1. we call `getTimestampFieldType` which returns `null`
2. some shards are unavailable for the `can_match` phase
3. we call `getTimestampFieldTypePatiently` to see for whether those shard failures can be ignored or not

**Review by fcofdez — APPROVED** [review]
LGTM, just some minor comments in the test 👍

File: x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java [hunk 2]
```diff
@@ -97,4 +102,53 @@ public void testTimestampRangeRecalculatedOnStalePrimaryAllocation() throws IOEx
...
+        }
+
+        for (final IndicesService indicesService : internalCluster().getInstances(IndicesService.class)) {
+            assertNull(indicesService.getTimestampFieldType(index));
+        }
+
+        assertAcked(client().execute(FreezeIndexAction.INSTANCE, new FreezeRequest("index")).actionGet());
+        ensureGreen("index");
+        for (final IndicesService indicesService : internalCluster().getInstances(IndicesService.class)) {
+            final PlainActionFuture<DateFieldMapper.DateFieldType> timestampFieldTypeFuture = new PlainActionFuture<>();
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java:137]: We have to pass `timestampFieldType` out of the `assertBusy()` block somehow. The following `assertTrue(isDone)` is technically unnecessary, we can't get to this point without having completed the future, but it's useful to document this fact.

File: x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java [hunk 3]
```diff
@@ -97,4 +102,53 @@ public void testTimestampRangeRecalculatedOnStalePrimaryAllocation() throws IOEx
...
+            assertBusy(() -> {
+                final DateFieldMapper.DateFieldType timestampFieldType = indicesService.getTimestampFieldType(index);
+                assertNotNull(timestampFieldType);
+                timestampFieldTypeFuture.onResponse(timestampFieldType);
+            });
+            assertTrue(timestampFieldTypeFuture.isDone());
+            assertThat(timestampFieldTypeFuture.get().dateTimeFormatter().locale().toString(), equalTo(locale));
```
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java:144]: Ok let me try and remember the month names in French to give this assertion some teeth 🇫🇷

File: server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java (hunk 1 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/indices/TimestampFieldMapperService.java:168]: Right, although in that case there's no particular reason to expect shards to be unavailable.

File: x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/frozen-indices/src/internalClusterTest/java/org/elasticsearch/index/engine/FrozenIndexIT.java:162]: Done in https://github.com/elastic/elasticsearch/pull/65873/commits/79ef7b032e6fe89ccac1b6c7360be6a5695432a5. Remembering the names wasn't the hard bit, it was working out that in French we write month names lower-case, with a trailing `.`, and sometimes use more than 3 letters.

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/bwc

**Comment by DaveCTurner** [comment]
@elasticmachine please run elasticsearch-ci/2

**Comment by DaveCTurner** [comment]
The backport failed in CI: https://gradle-enterprise.elastic.co/s/s3kleye5lwxds/console-log?task=:x-pack:plugin:frozen-indices:internalClusterTest

No idea why yet but I've reverted it from `7.x` for now.

**Comment by DaveCTurner** [comment]
I reinstated the backport with a few JDK8-specific tweaks in #65925.

**Pull request merged**